aip/general/0129.md

```diff
@@ -0,0 +1,71 @@
+---
+id: 129
+state: draft
+created: 2023-06-21
+placement:
+  category: resource-design
+  order: 65
+---
+
+# Server-modified Values and Defaults
+
+### State-driven clients (aka Convergent clients)
+Imperative clients rely on the result of an RPC call to determine success or failure of an operation. State-driven clients (including declarative clients) are clients that use the state of an existing resource to determine actions that should be taken to bring that resource to a desired state. To determine if a set of updates was successful a state-driven client compares server side resource attributes with client defined values. The comparison feature ensures accuracy of a creation or an update but it requires services to treat the client set fields as read-only and diligently preserve those values.
+
+### Permadiff
+Services that update client-set fields communicate to state-driven clients that the desired change has not taken effect, which results in the state-driven client to continuously call the service with updates. Permadiff is an infinite loop scenario where a client continuously attempts to update a resource field or a set of fields that the API service also modifies.
+
+## Guidance
+
+### Server Side Defaults
+There are instances where a service has to ensure that a value always exists even if the client chooses not to specify one. For example: a client may create a caching service instance without specifying the software version to be used for the service. Such a scenario requires the service to pick a version. The service designer has a number of ways to handle server side defaults.
+
+#### Dynamic Defaults
+When a default value of a field may change at any point in time such a field is considered to have a dynamic default. In general, dynamic fields can be: 
+* generated (UUID)
+* allocated (dynamic IP address)
+* assigned (most recent software package version)
+
+An attribute with a dynamic default ***must*** be expressed as two fields in the API:
+* a mutable field that can be optionally set by the user
+* an immutable `OUTPUT_ONLY` field that records the final value decided on by the API service
+
+Example:
+```proto
+message VirtualMachine {
+  …
+  string ip_address = 4;
+  string provisioned_ip_address = 5 [
```

Comment:
Good idea.

I would recommend "effective_" as a prefix for the output-only field, in general. There are some common cases of dynamic defaults for which we could recommend more specific alternatives, such as "provisioned_", "allocated_", or "assigned_" for assigned values.

Reply:
I like `effective_` too - that's certainly a convention we happen to use in the .NET client libraries (with nothing to do with protos normally) for "I've got inputs from a number of places, potentially including defaults, but this is what I'm really using".

Reply:
could that be a new section in this AIP? I'm fine with recommendations on naming as long as we don't make it a hard requirement. Or if it is a hard requirement, then we choose a single prefix for consistencies sake.

Also I would actually argue that it should be a suffix: this ensures that the input value and the output value are side-by-side when sorting them:

```
ip_address
ip_address_provisioned
```

Reply:
+1 for effective --- good for meaning "the value in effect" regardless of whether it was allocated, applied, generated, selected, or otherwise determined.

Re suffix - I do appreciate the sorting concern - though adjective-first may have precedence. e.g. create_time/update_time vs time_created/time_updated.

Reply:
Added a prefix naming suggestion as a sub-section on effective values. Happy to move that around if there's a better structure to be had